multitenancy/src/main/java/org/axonframework/extensions/multitenancy/components/deadletterqueue/MultiTenantDeadLetterQueue.java

```diff
@@ -0,0 +1,406 @@
+import org.axonframework.common.transaction.NoTransactionManager;
+import org.axonframework.eventhandling.EventMessage;
+import org.axonframework.extensions.multitenancy.TenantWrappedTransactionManager;
+import org.axonframework.extensions.multitenancy.components.MultiTenantAwareComponent;
+import org.axonframework.extensions.multitenancy.components.NoSuchTenantException;
+import org.axonframework.extensions.multitenancy.components.TargetTenantResolver;
+import org.axonframework.extensions.multitenancy.components.TenantDescriptor;
+import org.axonframework.messaging.Message;
+import org.axonframework.messaging.deadletter.DeadLetter;
+import org.axonframework.messaging.deadletter.DeadLetterQueueOverflowException;
+import org.axonframework.messaging.deadletter.EnqueueDecision;
+import org.axonframework.messaging.deadletter.NoSuchDeadLetterException;
+import org.axonframework.messaging.deadletter.SequencedDeadLetterQueue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+import java.util.function.UnaryOperator;
+import java.util.stream.Collectors;
+import java.util.stream.StreamSupport;
+
+import static org.axonframework.common.BuilderUtils.assertNonNull;
+
+/**
+ * Implementation of a {@link SequencedDeadLetterQueue} that is aware of the tenants in the application. This
+ * implementation will delegate the dead letter queue to the correct tenant segment based on the
+ * {@link TargetTenantResolver} provided.
+ *
+ * @author Stefan Dragisic
+ * @since 4.8.0
+ */
+public class MultiTenantDeadLetterQueue<M extends EventMessage<?>>
+        implements SequencedDeadLetterQueue<M>, MultiTenantAwareComponent {
+
+    private static final Logger logger = LoggerFactory.getLogger(MultiTenantDeadLetterQueue.class);
+
+    private final Set<TenantDescriptor> tenants = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    private final Map<TenantDescriptor, SequencedDeadLetterQueue<M>> tenantSegments = new ConcurrentHashMap<>();
+    private final TargetTenantResolver<M> targetTenantResolver;
+    private final String processingGroup;
+    private Supplier<SequencedDeadLetterQueue<M>> deadLetterQueueSupplier = () -> null;
+
+    /**
+     * Builder class to instantiate a {@link MultiTenantDeadLetterQueue}.
+     *
+     * @param builder the {@link Builder} used to instantiate a {@link MultiTenantDeadLetterQueue} instance.
+     */
+    public MultiTenantDeadLetterQueue(MultiTenantDeadLetterQueue.Builder<M> builder) {
+        builder.validate();
+        this.targetTenantResolver = builder.targetTenantResolver;
+        this.processingGroup = builder.processingGroup;
+    }
+
+    /**
+     * Instantiate a Builder to be able to create a {@link MultiTenantDeadLetterQueue}.
```

Comment:
Both the `TargetTenantResolver` and `processingGroup` are hard requirements for this Builder. Would be good to describe that on the JavaDoc, so that users can benefit from that.